common/src/main/java/org/opensearch/ml/common/utils/IndexUtils.java

```diff
@@ -73,34 +74,46 @@ public static String replacePlaceholders(String mapping) throws IOException {
             throw new IllegalArgumentException("Mapping cannot be null or empty");
         }
 
+        // Preload resources into memory to avoid redundant I/O
+        Map<String, String> loadedPlaceholders = new HashMap<>();
         for (Map.Entry<String, String> placeholder : MAPPING_PLACEHOLDERS.entrySet()) {
             URL url = IndexUtils.class.getClassLoader().getResource(placeholder.getValue());
             if (url == null) {
                 throw new IOException("Resource not found: " + placeholder.getValue());
             }
 
-            String placeholderMapping = Resources.toString(url, Charsets.UTF_8);
-            mapping = mapping.replace(placeholder.getKey(), placeholderMapping);
+            loadedPlaceholders.put(placeholder.getKey(), Resources.toString(url, Charsets.UTF_8));
         }
 
-        return mapping;
+        StringBuilder result = new StringBuilder(mapping);
+        for (Map.Entry<String, String> entry : loadedPlaceholders.entrySet()) {
+            String placeholder = entry.getKey();
+            String replacement = entry.getValue();
+
+            // Replace all occurrences of the placeholder
+            int index;
+            while ((index = result.indexOf(placeholder)) != -1) {
+                result.replace(index, index + placeholder.length(), replacement);
+            }
+        }
+
+        return result.toString();
     }
 
     /**
-        - Checks if mapping is a valid json
-        - Validates mapping against a schema found in mappings/schema.json
-        - Schema validates the following:
-            - Below fields are present:
-                - "_meta"
-                    - "_meta.schema_version"
-                - "properties"
-            - No additional fields at root level
-            - No additional fields in "_meta" object
-            - "properties" is an object type
-            - "_meta" is an object type
-            - "_meta_.schema_version" provided type is integer
-    
-        Note: Validation can be made more strict if a specific schema is defined for each index.
+     * Checks if mapping is a valid json
```

Comment:
[Nit] if you want to be fancy :D 

```
/**
     * Checks if the provided mapping is a valid JSON and validates it against a schema.
     *
     * <p>The schema is located at <code>mappings/schema.json</code> and enforces the following validations:</p>
     *
     * <ul>
     *   <li>Mandatory fields:
     *     <ul>
     *       <li><code>_meta</code></li>
     *       <li><code>_meta.schema_version</code></li>
     *       <li><code>properties</code></li>
     *     </ul>
     *   </li>
     *   <li>No additional fields are allowed at the root level.</li>
     *   <li>No additional fields are allowed in the <code>_meta</code> object.</li>
     *   <li><code>properties</code> must be an object type.</li>
     *   <li><code>_meta</code> must be an object type.</li>
     *   <li><code>_meta.schema_version</code> must be an integer.</li>
     * </ul>
     *
     * <p><strong>Note:</strong> Validation can be made stricter if a specific schema is defined for each index.</p>
     */
```

Reply:
<img width="420" alt="Screenshot 2024-12-27 at 3 04 50 PM" src="https://github.com/user-attachments/assets/fee9e4e5-3739-4ab7-8007-166994b4a374" />
It will look like this.